File: lab_guide_control/lab_guide_control/pid.py
# PID controller used for visitor-tracking control, mirrored by the ESP32 firmware implementation.

from dataclasses import dataclass, field


@dataclass
class PID:
    kp: float
    ki: float = 0.0
    kd: float = 0.0
    output_limit: float = float('inf')
    integral_limit: float = float('inf')

    _integral: float = field(default=0.0, init=False)
    _prev_error: float | None = field(default=None, init=False)

    def reset(self) -> None:
        self._integral = 0.0
        self._prev_error = None

    def update(self, error: float, dt: float) -> float:
        if dt <= 0.0:
            raise ValueError('dt must be positive')

        self._integral = _clamp(self._integral + error * dt, self.integral_limit)
        derivative = 0.0 if self._prev_error is None else (error - self._prev_error) / dt
        self._prev_error = error

        output = self.kp * error + self.ki * self._integral + self.kd * derivative
        limited = _clamp(output, self.output_limit)

        # Unwind the integrator when the output saturates, so it cannot keep growing while the actuator is pinned.
        if limited != output and self.ki != 0.0:
            self._integral -= (output - limited) / self.ki
            self._integral = _clamp(self._integral, self.integral_limit)

        return limited


def _clamp(value: float, limit: float) -> float:
    return max(-limit, min(limit, value))

File: firmware/esp32_pid/src/pid.cpp
// PID controller implementation for the tracking loop.

#include "pid.h"

namespace {

float clamp(float value, float limit) {
  if (value > limit) return limit;
  if (value < -limit) return -limit;
  return value;
}

}  // namespace

PID::PID(float kp, float ki, float kd, float outputLimit, float integralLimit)
    : kp_(kp),
      ki_(ki),
      kd_(kd),
      outputLimit_(outputLimit),
      integralLimit_(integralLimit),
      integral_(0.0f),
      prevError_(0.0f),
      hasPrevError_(false) {}

void PID::setGains(float kp, float ki, float kd) {
  kp_ = kp;
  ki_ = ki;
  kd_ = kd;
}

void PID::reset() {
  integral_ = 0.0f;
  prevError_ = 0.0f;
  hasPrevError_ = false;
}

float PID::update(float error, float dt) {
  if (dt <= 0.0f) {
    return 0.0f;
  }

  integral_ = clamp(integral_ + error * dt, integralLimit_);

  float derivative = hasPrevError_ ? (error - prevError_) / dt : 0.0f;
  prevError_ = error;
  hasPrevError_ = true;

  float output = kp_ * error + ki_ * integral_ + kd_ * derivative;
  float limited = clamp(output, outputLimit_);

  // Unwind the integrator when the output saturates, so it cannot keep growing while the actuator is pinned.
  if (limited != output && ki_ != 0.0f) {
    integral_ -= (output - limited) / ki_;
    integral_ = clamp(integral_, integralLimit_);
  }

  return limited;
}
